ziti/src/main/kotlin/io/netfoundry/ziti/net/nio/AsyncTLSChannel.kt

```diff
@@ -0,0 +1,507 @@
+        v{"continuing handshake: ${engine.handshakeStatus}"}
+        try {
+            when (engine.handshakeStatus) {
+                SSLEngineResult.HandshakeStatus.NEED_WRAP -> {
+                    val wrapped = ByteBuffer.allocate(32 * 1024)
+                    val res = engine.wrap(ByteBuffer.allocate(0), wrapped)
+                    wrapped.flip()
+                    hsWrite(wrapped, hs)
+                }
+
+                SSLEngineResult.HandshakeStatus.NEED_UNWRAP -> {
+                    val output = ByteBuffer.allocate(32 * 1024)
+                    while (hs.input.hasRemaining() && engine.handshakeStatus == SSLEngineResult.HandshakeStatus.NEED_UNWRAP) {
+                        val res = engine.unwrap(hs.input, output)
+                        if (res.bytesProduced() > 0) {
+                            hsWrite(output, hs)
+                            return
+                        }
+                        if (res.status == BUFFER_UNDERFLOW) {
+                            break;
+                        }
+                    }
+
+
+                    if (engine.handshakeStatus == SSLEngineResult.HandshakeStatus.NEED_UNWRAP) {
+                        hs.input.compact()
+                        transport.read(hs.input, hs, object : CompletionHandler<Int, Handhaker> {
+                            override fun completed(n: Int, r: Handhaker) {
+                                r.input.flip()
+                                try {
+                                    val res = engine.unwrap(r.input, output)
+
+                                    r.input.compact().flip()
+
+                                    output.flip()
+                                    if (res.bytesProduced() > 0) {
+                                        hsWrite(output, r)
+                                    } else {
+                                        continueHandshake(hs);
+                                    }
+                                } catch (sslex: SSLException) {
+                                    r.result.completeExceptionally(sslex)
+                                }
+                            }
+
+                            override fun failed(exc: Throwable?, r: Handhaker) {
+                                r.result.completeExceptionally(exc)
+                            }
+                        })
+                    } else {
+                        CompletableFuture.runAsync { continueHandshake(hs) }
+                    }
+                }
+                SSLEngineResult.HandshakeStatus.NEED_TASK -> {
+                    CompletableFuture.runAsync {
+                        engine.delegatedTask?.run()
+                        continueHandshake(hs)
+                    }
+                }
+
+                SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING -> {
+                    state = State.connected
+                    hs.result.complete(engine.session)
+                }
+
+                else -> TODO()
```

Comment:
anything else to add to TODO()? I don't know what the functino does but this just caught my eye